src/m365/spo/commands/tenant/tenant-applicationcustomizer-add.ts

```diff
@@ -0,0 +1,189 @@
+        }
+
+        return true;
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    try {
+      if (this.verbose) {
+        logger.logToStderr(`Adding application customizer with clientSideComponentId ${args.options.clientSideComponentId} and title ${args.options.title} to the tenant app catalog.`);
+      }
+      const appCatalogUrl = await this.getAppCatalogUrl();
+      if (this.verbose) {
+        logger.logToStderr(`Got tenant app catalog url: ${appCatalogUrl}`);
+      }
+
+      const solutionId = await this.getSolutionIdFromComponentManifestList(appCatalogUrl, args.options.clientSideComponentId, logger);
+      const solution = await this.getSolutionFromAppCatalog(appCatalogUrl, solutionId, logger);
+
+      if (!solution.ContainsTenantWideExtension) {
+        throw `The solution does not contain an extension that can be deployed to all sites. Make sure that you've entered the correct component Id.`;
+      }
+      else if (!solution.SkipFeatureDeployment) {
+        throw 'The solution has not been deployed to all sites. Make sure to deploy this solution to all sites.';
+      }
+      await this.addTenantWideExtension(appCatalogUrl, args.options, logger);
+    }
+    catch (err: any) {
+      this.handleRejectedODataJsonPromise(err);
+    }
+  }
+
+  private async getAppCatalogUrl(): Promise<string> {
+    const spoTenantAppCatalogUrlGetCommandOutput: CommandOutput = await Cli.executeCommandWithOutput(spoTenantAppCatalogUrlGetCommand as Command, { options: { output: 'text', _: [] } });
+    const appCatalogUrl: string | undefined = spoTenantAppCatalogUrlGetCommandOutput.stdout;
+    if (!appCatalogUrl) {
+      throw 'Cannot add tenant-wide application customizer as app catalog cannot be found';
+    }
+    return appCatalogUrl;
+  }
+
+  private async getSolutionIdFromComponentManifestList(appCatalogUrl: string, clientSideComponentId: string, logger: Logger): Promise<string> {
+    if (this.verbose) {
+      logger.logToStderr('Retrieving component manifest item from the ComponentManifests list on the app catalog site so that we get the solution id');
+    }
+
+    const camlQuery = `<View><ViewFields><FieldRef Name='ClientComponentId'></FieldRef><FieldRef Name='SolutionId'></FieldRef></ViewFields><Query><Where><Eq><FieldRef Name='ClientComponentId' /><Value Type='Guid'>${clientSideComponentId}</Value></Eq></Where></Query></View>`;
+    const commandOptions: spoListItemListCommandOptions = {
+      webUrl: appCatalogUrl,
+      listUrl: `${urlUtil.getServerRelativeSiteUrl(appCatalogUrl)}/Lists/ComponentManifests`,
+      camlQuery: camlQuery,
+      verbose: this.verbose,
+      debug: this.debug,
+      output: 'json'
+    };
+    const output = await Cli.executeCommandWithOutput(spoListItemListCommand as Command, { options: { ...commandOptions, _: [] } });
+    const outputParsed = JSON.parse(output.stdout);
```

Comment:
Not sure what you mean with this. Could you please clarify?

Reply:
Hi @MathijsVerbeeck!
I mean that on every command that you execute from code (`Cli.executeCommandWithOutput`), you should add a log statement that writes it's `stderr` stream:

`if (this.verbose) { logger.logToStderr(appGetOutput.stderr); }`

If you don't do this, the actual execution of the called command is like a black hole. Debug output is not coming through as well as any verbose statements. So if you add this line it will show in the console, which is what you would generally want. 

So please add these log statements below every `Cli.executeCommandWithOutput` function, try it out with `--verbose` and `--debug` to see how it looks and if possible you should align your own verbose statements, so that there are no similar verbose lines.

Do you see what I mean?

Reply:
Totally get it. Thanks for the clarification. Will change it right now!